File: sanket-dalvi/Day13-14/permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums: return []
        arr = []
        def perm(curr, remaining):
            if not remaining:
                arr.append(curr)
            for i in range(len(remaining)):
                perm(curr+[remaining[i]], remaining[:i]+remaining[i+1:])
        perm([], nums)
        return arr

# Test
print(Solution().permute([1,2,3]))

File: parasmittal/Day 13/Permutations.cpp
class Solution {
public:
    void util(vector<int>& arr,int l,int r, vector<vector<int>>& out){
        if(l==r){
            out.push_back(arr);
        }
        else{
            for(int i=l;i<=r;i++){
                swap(arr[i],arr[l]);
                util(arr,l+1,r,out);
                swap(arr[i],arr[l]);
            }
        }
    }
    vector<vector<int>> permute(vector<int>& arr) {
        vector<vector<int>> out;
        int n=arr.size();
        util(arr,0,n-1,out);
        return out;
    }
};